src/Microsoft.Health.Checkpoints/Storage/StorageCheckpointClient.cs

```diff
@@ -0,0 +1,143 @@
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Health.Checkpoints.Storage
+{
+    public class StorageCheckpointClient : ICheckpointClient
+    {
+        private BlobContainerClient _storageClient;
+        private ILogger<StorageCheckpointClient> _logger;
+        private const string _lastProcessedDateTime = "LastProcessedDateTime";
+        private const string _lastProcessedIdentifier = "LastProcessedIdentifier";
+
+        public StorageCheckpointClient(BlobContainerClient containerClient, ILogger<StorageCheckpointClient> logger)
+        {
+            EnsureArg.IsNotNull(containerClient);
+
+            _storageClient = containerClient;
+            _logger = logger;
+        }
+
+        public virtual async Task<ICheckpoint> GetCheckpointAsync(string partition, string checkpointIdentifier)
+        {
+            EnsureArg.IsNotNullOrEmpty(partition, nameof(partition));
+            EnsureArg.IsNotNullOrEmpty(checkpointIdentifier, nameof(checkpointIdentifier));
+
+            ICheckpoint checkpoint = new Checkpoint();
+
+            try
+            {
+                var resultSegment = _storageClient.GetBlobsAsync(
+                                        traits: BlobTraits.Metadata,
+                                        states: BlobStates.All,
+                                        prefix: $"{partition}/{checkpointIdentifier}",
+                                        cancellationToken: CancellationToken.None)
+                                        .AsPages();
+
+                await foreach (Page<BlobItem> blobPage in resultSegment)
+                {
+                    if (blobPage.Values.Count == 0)
+                    {
+                        _logger.LogWarning($"No blob found for identifier {checkpointIdentifier}");
+                    }
+
+                    if (blobPage.Values.Count > 1)
+                    {
+                        _logger.LogWarning($"Multiple blobs found for identifier {checkpointIdentifier}");
+                    }
+
+                    foreach (BlobItem blobItem in blobPage.Values)
+                    {
+                        DateTimeOffset lastEventTimestamp = DateTime.MinValue;
+
+                        if (blobItem.Metadata.TryGetValue(_lastProcessedDateTime, out var dt))
+                        {
+                            DateTimeOffset.TryParse(dt, null, DateTimeStyles.AssumeUniversal, out lastEventTimestamp);
+                            checkpoint.LastProcessedDateTime = lastEventTimestamp;
+                        }
+
+                        if (blobItem.Metadata.TryGetValue(_lastProcessedIdentifier, out var id))
+                        {
+                            checkpoint.LastProcessedIdentifier = id;
+                        }
+
+                        if (checkpoint.LastProcessedDateTime == DateTime.MinValue || checkpoint.LastProcessedIdentifier == null)
```

Comment:
Probably would make more sense to use `checkpoint.LastProcessedDateTime == default`. Also, we would generally use something like `string.IsNullOrEmpty` or `string.IsNullOrWhitespace` if they're appropriate here.